// angular2-collapsible/esm5/lib/services/collapsible-event.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
var CollapsibleEvent = /** @class */ (function () {
    function CollapsibleEvent() {
    }
    return CollapsibleEvent;
}());
export { CollapsibleEvent };
if (false) {
    /** @type {?} */
    CollapsibleEvent.prototype.type;
}
var CollapsibleEventService = /** @class */ (function () {
    function CollapsibleEventService() {
        this.toggleCollapsibleItemSource = new Subject();
        // Observable string streams
        this.toggleCollapsibleItem$ = this.toggleCollapsibleItemSource.asObservable();
    }
    // Toggle collapsible item
    /**
     * @param {?=} event
     * @return {?}
     */
    CollapsibleEventService.prototype.toggleCollapsibleItem = /**
     * @param {?=} event
     * @return {?}
     */
    function (event) {
        this.toggleCollapsibleItemSource.next(event);
    };
    /**
     * @return {?}
     */
    CollapsibleEventService.prototype.unsubscribe = /**
     * @return {?}
     */
    function () {
        this.toggleCollapsibleItemSource.unsubscribe();
    };
    CollapsibleEventService.decorators = [
        { type: Injectable }
    ];
    return CollapsibleEventService;
}());
export { CollapsibleEventService };
if (false) {
    /** @type {?} */
    CollapsibleEventService.prototype.type;
    /** @type {?} */
    CollapsibleEventService.prototype.toggleCollapsibleItemSource;
    /** @type {?} */
    CollapsibleEventService.prototype.toggleCollapsibleItem$;
}
