// angular2-collapsible/esm5/lib/services/collapsible.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
var CollapsibleService = /** @class */ (function () {
    function CollapsibleService() {
        this.headers = new Array();
        this.bodies = new Array();
    }
    /**
     * @return {?}
     */
    CollapsibleService.prototype.getType = /**
     * @return {?}
     */
    function () {
        return this.type;
    };
    /**
     * @param {?} type
     * @return {?}
     */
    CollapsibleService.prototype.setType = /**
     * @param {?} type
     * @return {?}
     */
    function (type) {
        this.type = type;
    };
    /**
     * @return {?}
     */
    CollapsibleService.prototype.getCollapsibleTable = /**
     * @return {?}
     */
    function () {
        return this.collapsibleTable;
    };
    /**
     * @param {?} table
     * @return {?}
     */
    CollapsibleService.prototype.setCollapsibleTable = /**
     * @param {?} table
     * @return {?}
     */
    function (table) {
        this.collapsibleTable = table;
    };
    /**
     * @param {?} header
     * @return {?}
     */
    CollapsibleService.prototype.addListHeader = /**
     * @param {?} header
     * @return {?}
     */
    function (header) {
        this.headers.push(header);
    };
    /**
     * @param {?} body
     * @return {?}
     */
    CollapsibleService.prototype.addListBody = /**
     * @param {?} body
     * @return {?}
     */
    function (body) {
        this.bodies.push(body);
    };
    /**
     * @return {?}
     */
    CollapsibleService.prototype.collapseAll = /**
     * @return {?}
     */
    function () {
        this.bodies.forEach(function (collapsibleBodyComponent) {
            // set 'expanded' properties of all the CollapsibleBodyComponents to 'false'
            collapsibleBodyComponent.expanded = false;
            collapsibleBodyComponent.expandedState = collapsibleBodyComponent.expanded.toString();
            // emit 'toggleState' event for all the CollapsibleBodyComponents
            collapsibleBodyComponent.toggleState.emit(false);
        });
    };
    CollapsibleService.decorators = [
        { type: Injectable }
    ];
    return CollapsibleService;
}());
export { CollapsibleService };
if (false) {
    /** @type {?} */
    CollapsibleService.prototype.type;
    /** @type {?} */
    CollapsibleService.prototype.collapsibleTable;
    /** @type {?} */
    CollapsibleService.prototype.headers;
    /** @type {?} */
    CollapsibleService.prototype.bodies;
}
