// angular2-collapsible/esm5/lib/services/collapsible-animations.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { trigger, state, style, transition, animate } from '@angular/animations';
var CollapsibleAnimationsService = /** @class */ (function () {
    function CollapsibleAnimationsService() {
    }
    /**
     * @param {?} triggerName
     * @return {?}
     */
    CollapsibleAnimationsService.collapsibleBodyAnimations = /**
     * @param {?} triggerName
     * @return {?}
     */
    function (triggerName) {
        return [
            trigger(triggerName, [
                state('*', style({
                    height: 0,
                    paddingTop: 0,
                    paddingBottom: 0,
                    borderBottom: 'none',
                    boxShadow: 'none',
                    overflow: 'hidden'
                })),
                state('true', style({
                    boxShadow: 'inset 0 4px 4px -2px rgb(208, 208, 208), inset 0 -3px 2px -1px rgb(208, 208, 208)',
                    overflow: 'inherit'
                })),
                transition('* => true', animate(CollapsibleAnimationsService.collapsibleBodyAnimationDuration + ' ' +
                    CollapsibleAnimationsService.easeInQuad)),
                transition('* => false', animate(CollapsibleAnimationsService.collapsibleBodyAnimationDuration + ' ' +
                    CollapsibleAnimationsService.easeOutQuad))
            ])
        ];
    };
    CollapsibleAnimationsService.collapsibleBodyAnimationDuration = '0.2s';
    CollapsibleAnimationsService.easeInQuad = 'cubic-bezier(0.55, 0.085, 0.68, 0.53)';
    CollapsibleAnimationsService.easeOutQuad = 'cubic-bezier(0.25, 0.46, 0.45, 0.94)';
    CollapsibleAnimationsService.collapsibleTableRowAnimationDuration = '0.2s';
    CollapsibleAnimationsService.decorators = [
        { type: Injectable }
    ];
    return CollapsibleAnimationsService;
}());
export { CollapsibleAnimationsService };
if (false) {
    /** @type {?} */
    CollapsibleAnimationsService.collapsibleBodyAnimationDuration;
    /** @type {?} */
    CollapsibleAnimationsService.easeInQuad;
    /** @type {?} */
    CollapsibleAnimationsService.easeOutQuad;
    /** @type {?} */
    CollapsibleAnimationsService.collapsibleTableRowAnimationDuration;
}
