// angular2-collapsible/esm5/lib/collapsible-table-row-detail/collapsible-table-row-detail.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, ViewChildren } from '@angular/core';
import { CollapsibleBodyComponent } from '../collapsible-body/collapsible-body.component';
import { CollapsibleService } from '../services/collapsible.service';
import { Subject } from 'rxjs';
var CollapsibleTableRowDetailComponent = /** @class */ (function () {
    function CollapsibleTableRowDetailComponent(collapsibleService) {
        this.collapsibleService = collapsibleService;
        this.subject = new Subject();
    }
    /**
     * @return {?}
     */
    CollapsibleTableRowDetailComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.subject.asObservable().subscribe(function () {
            _this.viewListBodies.forEach(function (item) {
                item.toggleCollapsibleItem();
            });
        });
    };
    /**
     * @return {?}
     */
    CollapsibleTableRowDetailComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // store list bodies in 'CollapsibleService'
        this.viewListBodies.forEach(function (item) {
            _this.collapsibleService.addListBody(item);
        });
    };
    CollapsibleTableRowDetailComponent.decorators = [
        { type: Component, args: [{
                    selector: 'collapsible-table-row-detail',
                    template: "\n        <td colspan=\"100%\">\n            <collapsible-list-item>\n                <collapsible-body>\n                    <ng-content></ng-content>\n                </collapsible-body>\n            </collapsible-list-item>\n        </td>\n    ",
                    styles: ["\n        :host {\n            display: table-row;\n        }\n\n        :host collapsible-body,\n        :host td {\n            padding: 0;\n        }\n    "]
                }] }
    ];
    /** @nocollapse */
    CollapsibleTableRowDetailComponent.ctorParameters = function () { return [
        { type: CollapsibleService }
    ]; };
    CollapsibleTableRowDetailComponent.propDecorators = {
        expanded: [{ type: Input }],
        viewListBodies: [{ type: ViewChildren, args: [CollapsibleBodyComponent,] }]
    };
    return CollapsibleTableRowDetailComponent;
}());
export { CollapsibleTableRowDetailComponent };
if (false) {
    /** @type {?} */
    CollapsibleTableRowDetailComponent.prototype.subject;
    /** @type {?} */
    CollapsibleTableRowDetailComponent.prototype.expanded;
    /** @type {?} */
    CollapsibleTableRowDetailComponent.prototype.viewListBodies;
    /** @type {?} */
    CollapsibleTableRowDetailComponent.prototype.collapsibleService;
}
