// angular2-collapsible/esm2015/lib/collapsible-table-row-detail/collapsible-table-row-detail.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, ViewChildren } from '@angular/core';
import { CollapsibleBodyComponent } from '../collapsible-body/collapsible-body.component';
import { CollapsibleService } from '../services/collapsible.service';
import { Subject } from 'rxjs';
export class CollapsibleTableRowDetailComponent {
    /**
     * @param {?} collapsibleService
     */
    constructor(collapsibleService) {
        this.collapsibleService = collapsibleService;
        this.subject = new Subject();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.subject.asObservable().subscribe(() => {
            this.viewListBodies.forEach((item) => {
                item.toggleCollapsibleItem();
            });
        });
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // store list bodies in 'CollapsibleService'
        this.viewListBodies.forEach((item) => {
            this.collapsibleService.addListBody(item);
        });
    }
}
CollapsibleTableRowDetailComponent.decorators = [
    { type: Component, args: [{
                selector: 'collapsible-table-row-detail',
                template: `
        <td colspan="100%">
            <collapsible-list-item>
                <collapsible-body>
                    <ng-content></ng-content>
                </collapsible-body>
            </collapsible-list-item>
        </td>
    `,
                styles: [`
        :host {
            display: table-row;
        }

        :host collapsible-body,
        :host td {
            padding: 0;
        }
    `]
            }] }
];
/** @nocollapse */
CollapsibleTableRowDetailComponent.ctorParameters = () => [
    { type: CollapsibleService }
];
CollapsibleTableRowDetailComponent.propDecorators = {
    expanded: [{ type: Input }],
    viewListBodies: [{ type: ViewChildren, args: [CollapsibleBodyComponent,] }]
};
if (false) {
    /** @type {?} */
    CollapsibleTableRowDetailComponent.prototype.subject;
    /** @type {?} */
    CollapsibleTableRowDetailComponent.prototype.expanded;
    /** @type {?} */
    CollapsibleTableRowDetailComponent.prototype.viewListBodies;
    /** @type {?} */
    CollapsibleTableRowDetailComponent.prototype.collapsibleService;
}
