// angular2-collapsible/esm2015/lib/collapsible-list-item/collapsible-list-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ContentChildren } from '@angular/core';
import { CollapsibleBodyComponent } from '../collapsible-body/collapsible-body.component';
import { CollapsibleService } from '../services/collapsible.service';
import { CollapsibleEventService } from '../services/collapsible-event.service';
export class CollapsibleListItemComponent {
    /**
     * @param {?} collapsibleService
     */
    constructor(collapsibleService) {
        this.collapsibleService = collapsibleService;
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        // store list bodies in 'CollapsibleService'
        this.contentListBodies.forEach((item) => {
            this.collapsibleService.addListBody(item);
        });
    }
}
CollapsibleListItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'collapsible-list-item',
                template: `
        <ng-content select="collapsible-header"></ng-content>
        <ng-content select="collapsible-body"></ng-content>
    `,
                providers: [CollapsibleEventService]
            }] }
];
/** @nocollapse */
CollapsibleListItemComponent.ctorParameters = () => [
    { type: CollapsibleService }
];
CollapsibleListItemComponent.propDecorators = {
    contentListBodies: [{ type: ContentChildren, args: [CollapsibleBodyComponent,] }]
};
if (false) {
    /** @type {?} */
    CollapsibleListItemComponent.prototype.contentListBodies;
    /** @type {?} */
    CollapsibleListItemComponent.prototype.collapsibleService;
}
