// angular2-collapsible/esm5/lib/collapsible-body/collapsible-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ElementRef, HostBinding, Output, EventEmitter } from '@angular/core';
import { CollapsibleService } from '../services/collapsible.service';
import { CollapsibleEventService } from '../services/collapsible-event.service';
import { CollapsibleAnimationsService } from '../services/collapsible-animations.service';
var CollapsibleBodyComponent = /** @class */ (function () {
    function CollapsibleBodyComponent(el, collapsibleService, eventService) {
        this.el = el;
        this.collapsibleService = collapsibleService;
        this.eventService = eventService;
        this.toggleState = new EventEmitter();
    }
    /**
     * @return {?}
     */
    CollapsibleBodyComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.eventService.toggleCollapsibleItem$.subscribe(function () {
            _this.toggleCollapsibleItem();
        });
    };
    /**
     * @return {?}
     */
    CollapsibleBodyComponent.prototype.hasContent = /**
     * @return {?}
     */
    function () {
        var e_1, _a;
        try {
            for (var _b = tslib_1.__values(this.el.nativeElement.childNodes), _c = _b.next(); !_c.done; _c = _b.next()) {
                var child = _c.value;
                if ((/** @type {?} */ (child)).tagName != null) {
                    return true;
                }
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return false;
    };
    /**
     * @return {?}
     */
    CollapsibleBodyComponent.prototype.toggleCollapsibleItem = /**
     * @return {?}
     */
    function () {
        // toggle body's state only if it has any childs
        if (this.hasContent()) {
            if (this.collapsibleService.getType() === 'accordion') {
                /** @type {?} */
                var tempExpanded = this.expanded;
                this.collapsibleService.collapseAll();
                this.expanded = tempExpanded;
            }
            this.expanded = !this.expanded;
            this.expandedState = this.expanded.toString();
            // emit 'toggleState' event
            this.toggleState.emit(this.expanded);
        }
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    CollapsibleBodyComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        var e_2, _a;
        if (this.hasContent()) {
            try {
                for (var _b = tslib_1.__values(Object.keys(changes)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    if (key === 'expanded' &&
                        changes["expanded"].currentValue != null) {
                        // this.expanded = changes.expanded.currentValue;
                        this.expandedState = this.expanded.toString();
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
    };
    // Makes sure we don't have a memory leak by destroying the
    // Subscription when our component is destroyed
    /**
     * @return {?}
     */
    CollapsibleBodyComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.eventService.unsubscribe();
    };
    CollapsibleBodyComponent.decorators = [
        { type: Component, args: [{
                    selector: 'collapsible-body',
                    template: "<ng-content></ng-content>",
                    animations: CollapsibleAnimationsService.collapsibleBodyAnimations('collapsibleBodyState'),
                    styles: ["\n        :host {\n            display: block;\n            border-bottom: 1px solid #ddd;\n            box-sizing: border-box;\n            padding: 2rem;\n        }\n\n        .side-nav :host {\n            padding: 0;\n        }\n\n        .side-nav :host,\n        .side-nav.fixed :host {\n            border: 0;\n            background-color: #fff;\n        }\n    "]
                }] }
    ];
    /** @nocollapse */
    CollapsibleBodyComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: CollapsibleService },
        { type: CollapsibleEventService }
    ]; };
    CollapsibleBodyComponent.propDecorators = {
        expandedState: [{ type: HostBinding, args: ['@collapsibleBodyState',] }],
        expanded: [{ type: Input }],
        toggleState: [{ type: Output }]
    };
    return CollapsibleBodyComponent;
}());
export { CollapsibleBodyComponent };
if (false) {
    /** @type {?} */
    CollapsibleBodyComponent.prototype.expandedState;
    /** @type {?} */
    CollapsibleBodyComponent.prototype.expanded;
    /** @type {?} */
    CollapsibleBodyComponent.prototype.toggleState;
    /** @type {?} */
    CollapsibleBodyComponent.prototype.el;
    /** @type {?} */
    CollapsibleBodyComponent.prototype.collapsibleService;
    /** @type {?} */
    CollapsibleBodyComponent.prototype.eventService;
}
