// angular2-collapsible/esm5/lib/collapsible-list-item/collapsible-list-item.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, ContentChildren } from '@angular/core';
import { CollapsibleBodyComponent } from '../collapsible-body/collapsible-body.component';
import { CollapsibleService } from '../services/collapsible.service';
import { CollapsibleEventService } from '../services/collapsible-event.service';
var CollapsibleListItemComponent = /** @class */ (function () {
    function CollapsibleListItemComponent(collapsibleService) {
        this.collapsibleService = collapsibleService;
    }
    /**
     * @return {?}
     */
    CollapsibleListItemComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // store list bodies in 'CollapsibleService'
        this.contentListBodies.forEach(function (item) {
            _this.collapsibleService.addListBody(item);
        });
    };
    CollapsibleListItemComponent.decorators = [
        { type: Component, args: [{
                    selector: 'collapsible-list-item',
                    template: "\n        <ng-content select=\"collapsible-header\"></ng-content>\n        <ng-content select=\"collapsible-body\"></ng-content>\n    ",
                    providers: [CollapsibleEventService]
                }] }
    ];
    /** @nocollapse */
    CollapsibleListItemComponent.ctorParameters = function () { return [
        { type: CollapsibleService }
    ]; };
    CollapsibleListItemComponent.propDecorators = {
        contentListBodies: [{ type: ContentChildren, args: [CollapsibleBodyComponent,] }]
    };
    return CollapsibleListItemComponent;
}());
export { CollapsibleListItemComponent };
if (false) {
    /** @type {?} */
    CollapsibleListItemComponent.prototype.contentListBodies;
    /** @type {?} */
    CollapsibleListItemComponent.prototype.collapsibleService;
}
