// angular2-collapsible/esm5/lib/collapsible.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CollapsibleListComponent } from './collapsible-list/collapsible-list.component';
import { CollapsibleListItemComponent } from './collapsible-list-item/collapsible-list-item.component';
import { CollapsibleHeaderComponent } from './collapsible-header/collapsible-header.component';
import { CollapsibleBodyComponent } from './collapsible-body/collapsible-body.component';
import { CollapsibleTableComponent } from './collapsible-table/collapsible-table.component';
import { CollapsibleTableRowComponent } from './collapsible-table-row/collapsible-table-row.component';
import { CollapsibleTableRowDetailComponent } from './collapsible-table-row-detail/collapsible-table-row-detail.component';
import { CollapsibleService } from './services/collapsible.service';
import { CollapsibleEventService } from './services/collapsible-event.service';
import { CollapsibleAnimationsService } from './services/collapsible-animations.service';
var CollapsibleModule = /** @class */ (function () {
    function CollapsibleModule() {
    }
    /**
     * @return {?}
     */
    CollapsibleModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: CollapsibleModule,
            providers: [CollapsibleModule]
        };
    };
    CollapsibleModule.decorators = [
        { type: NgModule, args: [{
                    imports: [
                        CommonModule
                    ],
                    declarations: [
                        CollapsibleListComponent,
                        CollapsibleListItemComponent,
                        CollapsibleHeaderComponent,
                        CollapsibleBodyComponent,
                        CollapsibleTableComponent,
                        CollapsibleTableRowComponent,
                        CollapsibleTableRowDetailComponent
                    ],
                    exports: [
                        CollapsibleListComponent,
                        CollapsibleListItemComponent,
                        CollapsibleHeaderComponent,
                        CollapsibleBodyComponent,
                        CollapsibleTableComponent,
                        CollapsibleTableRowComponent,
                        CollapsibleTableRowDetailComponent
                    ],
                    providers: [
                        CollapsibleService,
                        CollapsibleEventService,
                        CollapsibleAnimationsService
                    ]
                },] }
    ];
    return CollapsibleModule;
}());
export { CollapsibleModule };
