// angular2-collapsible/esm2015/lib/services/collapsible.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
export class CollapsibleService {
    constructor() {
        this.headers = new Array();
        this.bodies = new Array();
    }
    /**
     * @return {?}
     */
    getType() {
        return this.type;
    }
    /**
     * @param {?} type
     * @return {?}
     */
    setType(type) {
        this.type = type;
    }
    /**
     * @return {?}
     */
    getCollapsibleTable() {
        return this.collapsibleTable;
    }
    /**
     * @param {?} table
     * @return {?}
     */
    setCollapsibleTable(table) {
        this.collapsibleTable = table;
    }
    /**
     * @param {?} header
     * @return {?}
     */
    addListHeader(header) {
        this.headers.push(header);
    }
    /**
     * @param {?} body
     * @return {?}
     */
    addListBody(body) {
        this.bodies.push(body);
    }
    /**
     * @return {?}
     */
    collapseAll() {
        this.bodies.forEach((collapsibleBodyComponent) => {
            // set 'expanded' properties of all the CollapsibleBodyComponents to 'false'
            collapsibleBodyComponent.expanded = false;
            collapsibleBodyComponent.expandedState = collapsibleBodyComponent.expanded.toString();
            // emit 'toggleState' event for all the CollapsibleBodyComponents
            collapsibleBodyComponent.toggleState.emit(false);
        });
    }
}
CollapsibleService.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    CollapsibleService.prototype.type;
    /** @type {?} */
    CollapsibleService.prototype.collapsibleTable;
    /** @type {?} */
    CollapsibleService.prototype.headers;
    /** @type {?} */
    CollapsibleService.prototype.bodies;
}
