// angular2-collapsible/esm2015/lib/services/collapsible-event.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { Subject } from 'rxjs';
export class CollapsibleEvent {
}
if (false) {
    /** @type {?} */
    CollapsibleEvent.prototype.type;
}
export class CollapsibleEventService {
    constructor() {
        this.toggleCollapsibleItemSource = new Subject();
        // Observable string streams
        this.toggleCollapsibleItem$ = this.toggleCollapsibleItemSource.asObservable();
    }
    /**
     * @param {?=} event
     * @return {?}
     */
    toggleCollapsibleItem(event) {
        this.toggleCollapsibleItemSource.next(event);
    }
    /**
     * @return {?}
     */
    unsubscribe() {
        this.toggleCollapsibleItemSource.unsubscribe();
    }
}
CollapsibleEventService.decorators = [
    { type: Injectable }
];
if (false) {
    /** @type {?} */
    CollapsibleEventService.prototype.type;
    /** @type {?} */
    CollapsibleEventService.prototype.toggleCollapsibleItemSource;
    /** @type {?} */
    CollapsibleEventService.prototype.toggleCollapsibleItem$;
}
