// angular2-collapsible/esm5/lib/collapsible-header/collapsible-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, HostListener } from '@angular/core';
import { CollapsibleEventService } from '../services/collapsible-event.service';
var CollapsibleHeaderComponent = /** @class */ (function () {
    function CollapsibleHeaderComponent(eventService) {
        this.eventService = eventService;
    }
    /**
     * @return {?}
     */
    CollapsibleHeaderComponent.prototype.click = /**
     * @return {?}
     */
    function () {
        this.eventService.toggleCollapsibleItem();
    };
    CollapsibleHeaderComponent.decorators = [
        { type: Component, args: [{
                    selector: 'collapsible-header',
                    template: "<ng-content></ng-content>",
                    styles: ["\n        :host {\n            display: block;\n            cursor: pointer;\n            min-height: 3rem;\n            line-height: 3rem;\n            padding: 0 1rem;\n            background-color: #fff;\n            border-bottom: 1px solid #ddd;\n        }\n    "]
                }] }
    ];
    /** @nocollapse */
    CollapsibleHeaderComponent.ctorParameters = function () { return [
        { type: CollapsibleEventService }
    ]; };
    CollapsibleHeaderComponent.propDecorators = {
        click: [{ type: HostListener, args: ['click',] }]
    };
    return CollapsibleHeaderComponent;
}());
export { CollapsibleHeaderComponent };
if (false) {
    /** @type {?} */
    CollapsibleHeaderComponent.prototype.eventService;
}
