// angular2-collapsible/esm2015/lib/collapsible-header/collapsible-header.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, HostListener } from '@angular/core';
import { CollapsibleEventService } from '../services/collapsible-event.service';
export class CollapsibleHeaderComponent {
    /**
     * @param {?} eventService
     */
    constructor(eventService) {
        this.eventService = eventService;
    }
    /**
     * @return {?}
     */
    click() {
        this.eventService.toggleCollapsibleItem();
    }
}
CollapsibleHeaderComponent.decorators = [
    { type: Component, args: [{
                selector: 'collapsible-header',
                template: `<ng-content></ng-content>`,
                styles: [`
        :host {
            display: block;
            cursor: pointer;
            min-height: 3rem;
            line-height: 3rem;
            padding: 0 1rem;
            background-color: #fff;
            border-bottom: 1px solid #ddd;
        }
    `]
            }] }
];
/** @nocollapse */
CollapsibleHeaderComponent.ctorParameters = () => [
    { type: CollapsibleEventService }
];
CollapsibleHeaderComponent.propDecorators = {
    click: [{ type: HostListener, args: ['click',] }]
};
if (false) {
    /** @type {?} */
    CollapsibleHeaderComponent.prototype.eventService;
}
