// angular2-collapsible/esm2015/lib/collapsible-body/collapsible-body.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, ElementRef, HostBinding, Output, EventEmitter } from '@angular/core';
import { CollapsibleService } from '../services/collapsible.service';
import { CollapsibleEventService } from '../services/collapsible-event.service';
import { CollapsibleAnimationsService } from '../services/collapsible-animations.service';
export class CollapsibleBodyComponent {
    /**
     * @param {?} el
     * @param {?} collapsibleService
     * @param {?} eventService
     */
    constructor(el, collapsibleService, eventService) {
        this.el = el;
        this.collapsibleService = collapsibleService;
        this.eventService = eventService;
        this.toggleState = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.eventService.toggleCollapsibleItem$.subscribe(() => {
            this.toggleCollapsibleItem();
        });
    }
    /**
     * @return {?}
     */
    hasContent() {
        for (const child of this.el.nativeElement.childNodes) {
            if ((/** @type {?} */ (child)).tagName != null) {
                return true;
            }
        }
        return false;
    }
    /**
     * @return {?}
     */
    toggleCollapsibleItem() {
        // toggle body's state only if it has any childs
        if (this.hasContent()) {
            if (this.collapsibleService.getType() === 'accordion') {
                /** @type {?} */
                const tempExpanded = this.expanded;
                this.collapsibleService.collapseAll();
                this.expanded = tempExpanded;
            }
            this.expanded = !this.expanded;
            this.expandedState = this.expanded.toString();
            // emit 'toggleState' event
            this.toggleState.emit(this.expanded);
        }
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (this.hasContent()) {
            for (const key of Object.keys(changes)) {
                if (key === 'expanded' &&
                    changes["expanded"].currentValue != null) {
                    // this.expanded = changes.expanded.currentValue;
                    this.expandedState = this.expanded.toString();
                }
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.eventService.unsubscribe();
    }
}
CollapsibleBodyComponent.decorators = [
    { type: Component, args: [{
                selector: 'collapsible-body',
                template: `<ng-content></ng-content>`,
                animations: CollapsibleAnimationsService.collapsibleBodyAnimations('collapsibleBodyState'),
                styles: [`
        :host {
            display: block;
            border-bottom: 1px solid #ddd;
            box-sizing: border-box;
            padding: 2rem;
        }

        .side-nav :host {
            padding: 0;
        }

        .side-nav :host,
        .side-nav.fixed :host {
            border: 0;
            background-color: #fff;
        }
    `]
            }] }
];
/** @nocollapse */
CollapsibleBodyComponent.ctorParameters = () => [
    { type: ElementRef },
    { type: CollapsibleService },
    { type: CollapsibleEventService }
];
CollapsibleBodyComponent.propDecorators = {
    expandedState: [{ type: HostBinding, args: ['@collapsibleBodyState',] }],
    expanded: [{ type: Input }],
    toggleState: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CollapsibleBodyComponent.prototype.expandedState;
    /** @type {?} */
    CollapsibleBodyComponent.prototype.expanded;
    /** @type {?} */
    CollapsibleBodyComponent.prototype.toggleState;
    /** @type {?} */
    CollapsibleBodyComponent.prototype.el;
    /** @type {?} */
    CollapsibleBodyComponent.prototype.collapsibleService;
    /** @type {?} */
    CollapsibleBodyComponent.prototype.eventService;
}
