// angular2-collapsible/esm5/lib/collapsible-list/collapsible-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Component, Input, ContentChildren } from '@angular/core';
import { CollapsibleListItemComponent } from '../collapsible-list-item/collapsible-list-item.component';
import { CollapsibleService } from '../services/collapsible.service';
var CollapsibleListComponent = /** @class */ (function () {
    function CollapsibleListComponent(collapsibleService) {
        this.collapsibleService = collapsibleService;
        // component options
        //
        // describes the type of the collapsible list: 'accordion' or 'expandable'
        this.type = 'accordion';
    }
    /**
     * @return {?}
     */
    CollapsibleListComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.collapsibleService.setType(this.type);
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    CollapsibleListComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        for (var change in changes) {
            if (change === 'type') {
                this.type = changes["type"].currentValue;
                this.collapsibleService.setType(this.type);
            }
        }
    };
    CollapsibleListComponent.decorators = [
        { type: Component, args: [{
                    selector: 'collapsible-list',
                    template: "<ng-content select=\"collapsible-list-item\"></ng-content>",
                    providers: [CollapsibleService],
                    styles: ["\n    :host {\n      display: block;\n      margin: 0.5rem 0 1rem 0;\n      border-top: 1px solid #ddd;\n      border-right: 1px solid #ddd;\n      border-left: 1px solid #ddd;\n      box-shadow:\n        0 2px 2px 0 rgba(0, 0, 0, 0.14),\n        0 1px 5px 0 rgba(0, 0, 0, 0.12),\n        0 3px 1px -2px rgba(0, 0, 0, 0.2);\n    }\n\n    .side-nav :host,\n    .side-nav.fixed :host {\n      border: none;\n      box-shadow: none;\n    }\n\n    :host(.popout) {\n      border: none;\n      box-shadow: none;\n    }\n\n    :host(.popout) > li {\n      box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.16), 0 2px 10px 0 rgba(0, 0, 0, 0.12);\n      margin: 0 24px;\n      transition: margin 0.35s cubic-bezier(0.25, 0.46, 0.45, 0.94);\n    }\n\n    .side-nav :host {\n      margin: 0;\n    }\n  "]
                }] }
    ];
    /** @nocollapse */
    CollapsibleListComponent.ctorParameters = function () { return [
        { type: CollapsibleService }
    ]; };
    CollapsibleListComponent.propDecorators = {
        type: [{ type: Input }],
        contentListItems: [{ type: ContentChildren, args: [CollapsibleListItemComponent,] }]
    };
    return CollapsibleListComponent;
}());
export { CollapsibleListComponent };
if (false) {
    /** @type {?} */
    CollapsibleListComponent.prototype.type;
    /** @type {?} */
    CollapsibleListComponent.prototype.contentListItems;
    /** @type {?} */
    CollapsibleListComponent.prototype.collapsibleService;
}
